src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -1068,75 +1068,35 @@ void ReferenceProcessor::preclean_discovered_references(BoolObjectClosure* is_al
-        return;
-      }
-    }
-    log_reflist("SoftRef after: ", _discoveredSoftRefs, _max_num_queues);
-  }
+  ReferenceType ref_type_arr[4] = { REF_SOFT, REF_WEAK, REF_FINAL, REF_PHANTOM };
+  size_t ref_count_arr[4] = {};
 
-  // Weak references
-  {
-    GCTraceTime(Debug, gc, ref) tm("Preclean WeakReferences", gc_timer);
-    log_reflist("WeakRef before: ", _discoveredWeakRefs, _max_num_queues);
-    for (uint i = 0; i < _max_num_queues; i++) {
-      if (yield->should_return()) {
-        return;
-      }
-      if (preclean_discovered_reflist(_discoveredWeakRefs[i], is_alive,
-                                      enqueue, yield)) {
-        log_reflist("WeakRef abort: ", _discoveredWeakRefs, _max_num_queues);
-        return;
-      }
+  if (discovery_is_mt()) {
+    for (int i = 0; i < 4; ++i) {
+      ReferenceType ref_type = ref_type_arr[i];
+      DiscoveredList* list = get_discovered_list(ref_type);
+      ref_count_arr[i] = list->length();
+      preclean_discovered_reflist(*list, is_alive, enqueue, yield);
     }
-    log_reflist("WeakRef after: ", _discoveredWeakRefs, _max_num_queues);
-  }
-
-  // Final references
-  {
-    GCTraceTime(Debug, gc, ref) tm("Preclean FinalReferences", gc_timer);
-    log_reflist("FinalRef before: ", _discoveredFinalRefs, _max_num_queues);
-    for (uint i = 0; i < _max_num_queues; i++) {
-      if (yield->should_return()) {
-        return;
-      }
-      if (preclean_discovered_reflist(_discoveredFinalRefs[i], is_alive,
-                                      enqueue, yield)) {
-        log_reflist("FinalRef abort: ", _discoveredFinalRefs, _max_num_queues);
-        return;
+  } else {
+    for (int i = 0; i < 4; ++i) {
+      ReferenceType ref_type = ref_type_arr[i];
+      // When discovery is *not* multi-threaded, discovered refs are stored in
+      // list[0.._num_queues-1]. Loop _num_queues times to cover all lists.
+      for (uint j = 0; j < _num_queues; ++j) {
+        DiscoveredList* list = get_discovered_list(ref_type);
+        ref_count_arr[i] += list->length();
+        preclean_discovered_reflist(*list, is_alive, enqueue, yield);
       }
     }
-    log_reflist("FinalRef after: ", _discoveredFinalRefs, _max_num_queues);
   }
 
-  // Phantom references
-  {
-    GCTraceTime(Debug, gc, ref) tm("Preclean PhantomReferences", gc_timer);
-    log_reflist("PhantomRef before: ", _discoveredPhantomRefs, _max_num_queues);
-    for (uint i = 0; i < _max_num_queues; i++) {
-      if (yield->should_return()) {
-        return;
-      }
-      if (preclean_discovered_reflist(_discoveredPhantomRefs[i], is_alive,
-                                      enqueue, yield)) {
-        log_reflist("PhantomRef abort: ", _discoveredPhantomRefs, _max_num_queues);
-        return;
-      }
-    }
-    log_reflist("PhantomRef after: ", _discoveredPhantomRefs, _max_num_queues);
-  }
+  uint worker_id = WorkerThread::current()->id();
```

Comment:
This logging is potentially far more verbose than previously, since there will now be a line for each concurrent marking thread, rather than one line for the previously single-threaded precleaning.  We have mechanisms for collecting and reporting timing and work units for parallel activities that should be used here.

Reply:
The new logs are on `trace` level, just FYI. How about I address all logging related issues (from Thomas and Stefan as well) in a follow-up PR?

Reply:
I know this is trace logging, but that doesn't make the verbosity or content good.  Consider a machine with a large amount of concurrency; one might be interested in any of the totals/min/max/avg, and having to extract that manually is going to be annoying.  I'm okay with it being a followup though.

Reply:
> one might be interested in any of the totals/min/max/avg

Thomas suggested this as well. I will address this in a followup PR.